File: 207.course-schedule.py
#
# @lc app=leetcode id=207 lang=python3
#
# [207] Course Schedule
#

# @lc code=start
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:


        #create dict: course: [prereqs]
        pre_dict = {i:[] for i in range(numCourses)}

        for i in prerequisites:
            c = i[0]
            p = i[1]
            pre_dict[c].append(p)

        #go trough dict with a bfs

        visit_list = []

        def dfs(c):

            #basecases
            if c  in visit_list:
                return False

            if pre_dict[c] == []:
                return True

            visit_list.append(c)

            for p in pre_dict[c]:
                if not dfs(p): return False
            
            visit_list.remove(c)
            pre_dict[c] = []
            return True

        
        for c in range(numCourses):
            if not dfs(c): return False

        return True




# @lc code=end

File: 207.course-schedule.cpp
/*
 * @lc app=leetcode id=207 lang=cpp
 *
 * [207] Course Schedule
 */

// @lc code=start
class Solution
{
public:
    bool dfs(int course, unordered_map<int, vector<int>> &m, unordered_set<int> &visited){

        //check if visites
        if (visited.find(course) != visited.end())
        {
            return false;
        }
        

        //check if no pre req
        if (m[course].empty())
        {
            return true;
        }

        //check dfs
        visited.insert(course);
        for (int i = 0; i < m[course].size(); i++)
        {
            int nextCourse = m[course][i];
            if (!dfs(nextCourse, m, visited))
            {
                return false;
            }
        }
        m[course].clear();
        visited.erase(course);
        return true;
    }

        bool canFinish(int numCourses, vector<vector<int>> &prerequisites)
    {
        // map each course to prereq list
        unordered_map<int, vector<int>> m;
        for (int i = 0; i < prerequisites.size(); i++)
        {
            m[prerequisites[i][0]].push_back(prerequisites[i][1]);
        }

        // all courses along current DFS path
        unordered_set<int> visited;

        for (int course = 0; course < numCourses; course++)
        {
            if (!dfs(course, m, visited))
            {
                return false;
            }
        }
        return true;
    }
};
// @lc code=end
